shared-bindings/wiznet/__init__.c

```diff
@@ -0,0 +1,526 @@
+        1, // SECOND PHASE TRANSITION
+        8 // 8 BITS
+    );
+
+    common_hal_digitalio_digitalinout_switch_to_output(&wiznet5k_obj.cs, 1, DRIVE_MODE_PUSH_PULL);
+    common_hal_digitalio_digitalinout_switch_to_output(&wiznet5k_obj.rst, 1, DRIVE_MODE_PUSH_PULL); 
+
+    common_hal_digitalio_digitalinout_set_value(&wiznet5k_obj.rst, 0);
+    mp_hal_delay_us(10); // datasheet says 2us
+    common_hal_digitalio_digitalinout_set_value(&wiznet5k_obj.rst, 1);
+    mp_hal_delay_ms(160); // datasheet says 150ms
+
+    reg_wizchip_cris_cbfunc(wiz_cris_enter, wiz_cris_exit);
+    reg_wizchip_cs_cbfunc(wiz_cs_select, wiz_cs_deselect);
+    reg_wizchip_spi_cbfunc(wiz_spi_read, wiz_spi_write);
+
+    uint8_t sn_size[16] = {2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2}; // 2k buffer for each socket
+    ctlwizchip(CW_INIT_WIZCHIP, sn_size);
+
+    // set some sensible default values; they are configurable using ifconfig method
+    wiz_NetInfo netinfo = {
+        .ip = {192, 168, 0, 18},
+        .sn = {255, 255, 255, 0},
+        .gw = {192, 168, 0, 1},
+        .dns = {8, 8, 8, 8}, // Google public DNS
+        .dhcp = NETINFO_STATIC,
+    };
+    create_random_mac_address(netinfo.mac);
+    ctlnetwork(CN_SET_NETINFO, (void*)&netinfo);
+
+    // seems we need a small delay after init
+    mp_hal_delay_ms(250);
+
+    // register with network module
+    network_module_register_nic(&wiznet5k_obj);
+
+    // return wiznet5k object
+    return &wiznet5k_obj;
+}
+
+/// \method regs()
+/// Dump WIZNET5K registers.
+STATIC mp_obj_t wiznet5k_regs(mp_obj_t self_in) {
+    //wiznet5k_obj_t *self = self_in;
+    printf("Wiz CREG:");
+    for (int i = 0; i < 0x50; ++i) {
+        if (i % 16 == 0) {
+            printf("\n  %04x:", i);
+        }
+        #if MICROPY_PY_WIZNET5K == 5200
+        uint32_t reg = i;
+        #else
+        uint32_t reg = _W5500_IO_BASE_ | i << 8;
+        #endif
+        printf(" %02x", WIZCHIP_READ(reg));
+    }
+    for (int sn = 0; sn < 4; ++sn) {
+        printf("\nWiz SREG[%d]:", sn);
+        for (int i = 0; i < 0x30; ++i) {
+            if (i % 16 == 0) {
+                printf("\n  %04x:", i);
+            }
+            #if MICROPY_PY_WIZNET5K == 5200
+            uint32_t reg = WIZCHIP_SREG_ADDR(sn, i);
+            #else
+            uint32_t reg = _W5500_IO_BASE_ | i << 8 | WIZCHIP_SREG_BLOCK(sn) << 3;
+            #endif
+            printf(" %02x", WIZCHIP_READ(reg));
+        }
+    }
+    printf("\n");
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(wiznet5k_regs_obj, wiznet5k_regs);
+
+STATIC mp_obj_t wiznet5k_isconnected(mp_obj_t self_in) {
+    (void)self_in;
+    return mp_obj_new_bool(wizphy_getphylink() == PHY_LINK_ON);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(wiznet5k_isconnected_obj, wiznet5k_isconnected);
+
+/// \method ifconfig([(ip, subnet, gateway, dns)])
+/// Get/set IP address, subnet mask, gateway and DNS.
+STATIC mp_obj_t wiznet5k_ifconfig(size_t n_args, const mp_obj_t *args) {
```

Comment:
How about adding properties for each of these four attributes? Do they all need to be set at the once? If not, they can be read/write. Otherwise they can be read-only.

If they all need to be set at once, can we switch to a method that takes kwargs so each ip is labeled with what it is?

Reply:
Yeah, they're all set together.  It'd make sense for these to be kwargs I agree, it does make the code more readable.